csrc/ir/nodes.cpp

```diff
@@ -389,9 +389,81 @@ UnaryOp::UnaryOp(IrBuilderPasskey passkey, UnaryOpType type, Val* out, Val* in)
 
 std::vector<PolymorphicValue> UnaryOp::evaluate(
     const ExpressionEvaluator& ee,
-    const std::vector<PolymorphicValue>& inputs) const {
+    std::unordered_map<const Val*, PolymorphicValue>& known_values) const {
   using namespace PolymorphicValue_functions;
-  const auto& in = inputs.at(0);
+
+  // If the UnaryOp is CastOp, check if the preceding pattern of
+  // operators matches with matmul (MmaOp(Broadcast (A), Broadcast(B)) -> Cast)
+  // or matmul + bias (BinaryOp::Add (MmaOp(Broadcast (A), Broadcast(B),
+  // Broadcast(bias)) -> Cast) If not, evaluate UnaryOp::CastOp along with the
+  // other types by evaluating the immediate input.
+
+  auto has_matmul_and_bias = [](Val* in) -> bool {
+    if (auto* binary = dynamic_cast<BinaryOp*>(in->definition())) {
+      if (binary->getBinaryOpType() == BinaryOpType::Add) {
+        if (binary->input(0)->definition()->isA<MmaOp>()) {
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  if ((getUnaryOpType() == UnaryOpType::Cast) &&
+      input(0)->definition() != nullptr) {
+    // Case 1: MmaOp + Cast (Matmul)
+    if (auto* mma = dynamic_cast<MmaOp*>(input(0)->definition())) {
+      MmaOpUtils::verifyMmaOpForEvaluation(mma, out()->getDataType().value());
+
+      std::vector<at::Tensor> mma_inputs;
+      for (Val* inp : mma->inputs()) {
+        auto eval_i = ee.evaluate(inp, known_values);
+        mma_inputs.push_back(eval_i.as<at::Tensor>());
+      }
+
+      const auto a = mma_inputs.at(0).squeeze(-1);
+      const auto b = mma_inputs.at(1).squeeze(0);
+
+      // After removing the broadcast dimensions, the format should be
+      // [M, K] x [K, N] compatible with aten::matmul format.
+      return {a.matmul(b)};
+    }
+
+    // Case 2: Matmul + Bias
+    if (has_matmul_and_bias(input(0))) {
+      BinaryOp* binary = input(0)->definition()->as<BinaryOp>();
+      MmaOp* mma = binary->input(0)->definition()->as<MmaOp>();
+
+      MmaOpUtils::verifyMmaOpForEvaluation(mma, out()->getDataType().value());
+      MmaOpUtils::verifyBiasForEvaluation(
+          binary->input(1), out()->getDataType().value());
+
+      // BinaryOp <- Broadcast <- CastOp <- Bias
+      const Val* bias =
+          binary->input(1)->definition()->input(0)->definition()->input(0);
```

Comment:
Is this safe? has_matmul_and_bias doesn't seem to check the bias side.

Reply:
```
 MmaOpUtils::verifyMmaOpForEvaluation(mma, out()->getDataType().value());
      MmaOpUtils::verifyBiasForEvaluation(
          binary->input(1), out()->getDataType().value());
```

checks that the necessary conditions for the MmaOp and Bias are true.